Add revokeVotes method to Governance contract.

deepakhb2 opened the pull request:
### Description

Currently there is no option to revoke the votes for an account so that locked gold can be released.
Added revokeVotes method to revoke all the votes from referendum proposals for sender account.

### Other changes


### Tested

_An explanation of how the changes were tested or an explanation as to why they don't need to be._

### Related issues

- Fixes #7562 

### Backwards compatibility

Added new method so it is backward compatible.

### Documentation

N/A

deepakhb2 committed:
Add revokeVotes method to Governance contract.
Add unit tests for testing revokeVotes method.

deepakhb2 commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -648,6 +650,29 @@ contract Governance is
+      );
+      require(stage == Proposals.Stage.Referendum, "Incorrect proposal state");
+      proposal.updateVote(voteRecord.weight, 0, voteRecord.value, Proposals.VoteValue.None);
+      proposal.networkWeight = proposal.networkWeight - voteRecord.weight;
```
@yorhodes How to reduce the networkWeight? I am not sure how to calculate the networWeight of a proposal.
cc @nambrot

yorhodes replied:
the `networkWeight` isn't actually reduced until the voter unlocks
this should be updated as it is elsewhere
```suggestion
proposal.networkWeight = getLockedGold().getTotalLockedGold();
```

deepakhb2 committed:
Fix lint errors.

deepakhb2 committed:
Fix the lint errors in Governance contract.

yorhodes commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -148,6 +148,8 @@ contract Governance is
     uint256 weight
   );
 
+  event AccountProposalsVoteRevoked(address indexed account);
```
```suggestion
  event ProposalVoteRevoked(
    uint256 indexed proposalId,
    address indexed account,
    uint256 value,
    uint256 weight
  );
```

for consistency with `ProposalUnvoteRevoked` and `ProposalVoted`
We should be able to tell from these events the canonical vote tallies on a specific proposal and voter

yorhodes commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -648,6 +650,30 @@ contract Governance is
+      voter.referendumVotes[i] = VoteRecord(Proposals.VoteValue.None, voteRecord.proposalId, 0);
+    }
+    voter.mostRecentReferendumProposal = 0;
+    emit AccountProposalsVoteRevoked(account);
```
emit this for all unvoted proposals

yorhodes commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -648,6 +650,30 @@ contract Governance is
+  function revokeVotes() external nonReentrant returns (bool) {
+    address account = getAccounts().voteSignerToAccount(msg.sender);
+    Voter storage voter = voters[account];
+    for (uint256 i = 0; i < dequeued.length; i = i.add(1)) {
```
how about `dequeueIndex` instead of `i`?

yorhodes commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -648,6 +650,30 @@ contract Governance is
+      require(stage == Proposals.Stage.Referendum, "Incorrect proposal state");
+      proposal.updateVote(voteRecord.weight, 0, voteRecord.value, Proposals.VoteValue.None);
+      proposal.networkWeight = proposal.networkWeight - voteRecord.weight;
+      voter.referendumVotes[i] = VoteRecord(Proposals.VoteValue.None, voteRecord.proposalId, 0);
```
imo we should garbage collect here
```suggestion
      delete voter.referendumVotes[i];
```

yorhodes commented on packages/protocol/test/governance/network/governance.ts:
```diff
@@ -1500,6 +1500,113 @@ contract('Governance', (accounts: string[]) => {
+    it('should return true', async () => {
+      const success = await governance.revokeVotes.call()
+      assert.isTrue(success)
+    })
```
I'm not sure how useful this test is

deepakhb2 replied:
This is to verify that user knows that all votes are revoked by function returning true.

yorhodes replied:
`call` doesn't alter state but I see your point
not sure how familiar you are with the web3 landscape but this might be a good resource https://bitsofco.de/calling-smart-contract-functions-using-web3-js-call-vs-send/

deepakhb2 replied:
I have removed this test case as it will not modify the state. I wasn't aware of it and I copied it from vote tests.

yorhodes commented on packages/protocol/test/governance/network/governance.ts:
```diff
@@ -1500,6 +1500,113 @@ contract('Governance', (accounts: string[]) => {
+        assert.isFalse(voting)
+      })
+    })
+  })
```
if we emit events for each unvoted proposal, would love an additional unit test here for the two proposals `describe`

yorhodes reviewed (changes requested):
great start!

deepakhb2 committed:
Merge branch 'master' of github.com:celo-org/celo-monorepo into issue/7562-revoke-votes

deepakhb2 committed:
Fix the networkweight calculation.
Emmit Proposal vote revoked event for each unvoted proposal.
Upate unit tests accordingly.

deepakhb2 committed:
Merge branch 'issue/7562-revoke-votes' of github.com:deepakhb2/celo-monorepo into issue/7562-revoke-votes

deepakhb2 committed:
Merge branch 'master' of github.com:celo-org/celo-monorepo into issue/7562-revoke-votes

deepakhb2 committed:
Fix the lint error.

deepakhb2 committed:
Fix lint errors in Governance.sol contract.

deepakhb2 committed:
Ignore prettier error for a line in Governance contract.

yorhodes commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -657,20 +662,30 @@ contract Governance is
-      proposal.updateVote(voteRecord.weight, 0, voteRecord.value, Proposals.VoteValue.None);
-      proposal.networkWeight = proposal.networkWeight - voteRecord.weight;
-      voter.referendumVotes[i] = VoteRecord(Proposals.VoteValue.None, voteRecord.proposalId, 0);
+      uint256 weight = voteRecord.weight;
```
I don't feel super strongly about this but I think this variable unnecessarily expands the memory stack
we might want to just reference via `voteRecord.weight` throughout

yorhodes replied:
same for `voteRecord.value`

deepakhb2 replied:
Yeah, even I thought so but for some reason I had to do this. I will revert this back. thanks again.

yorhodes commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -657,20 +662,30 @@ contract Governance is
+        Proposals.VoteValue.None,
+        voteRecord.proposalId,
+        0
+      );
```
we can just remove this because of the `delete`
```suggestion
```

deepakhb2 replied:
I should have noticed this. Thank you.

yorhodes commented on packages/protocol/test/governance/network/governance.ts:
```diff
@@ -1604,6 +1596,11 @@ contract('Governance', (accounts: string[]) => {
         assert.isFalse(voting)
       })
+
+      it('should emit the ProposalVoteRevoked event 3 times', async () => {
```
nice!

yorhodes reviewed (approved):
nice work!
please note my comments before merging

deepakhb2 committed:
Remove code that is not required anymore.

yorhodes commented:
https://app.circleci.com/pipelines/github/celo-org/celo-monorepo/34421/workflows/4cfd72b5-c907-41ab-b3ac-72b681bc28ad/jobs/442800/parallel-runs/0/steps/0-104
is the same deal as your other PR @deepakhb2 
we need to make `Governance` be `InitializableV2` and in this case bump the `minor` version in `getVersionNumber` because we added a function to the bytecode

deepakhb2 committed:
Inherit from InitializabeV2 in Governance contract.

deepakhb2 committed:
Merge branch 'master' of github.com:celo-org/celo-monorepo into issue/7562-revoke-votes

deepakhb2 committed:
Update the patch version.

deepakhb2 committed:
Remove the duplicate code and test that was not required.

deepakhb2 committed:
Update version number.

deepakhb2 committed:
Update version number.

yorhodes commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -200,9 +200,15 @@ contract Governance is
    */
   function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
-    return (1, 2, 0, 3);
+    return (1, 2, 1, 2);
```
```suggestion
    return (1, 2, 1, 0);
```
the minor bump supercedes any patch change

yorhodes replied:
I'm actually not sure why this passed in CI...
@nambrot

deepakhb2 replied:
return (1, 2, 1, 0) is correct right? When minor is updated patch becomes 0. This is as you suggested in your previous message. I had to make multiple commits to fix this.

yorhodes replied:
yes, thank you!

nambrot replied:
@yorhodes sounds like a ticket to me :P

yorhodes reviewed (approved):
looks great!
